packages/flutter/lib/src/cupertino/sliding_segmented_control.dart

```diff
@@ -0,0 +1,1037 @@
+/// control cease to be selected.
+///
+/// A segmented control can feature any [Widget] as one of the values in its
+/// [Map] of [children]. The type T is the type of the keys used
+/// to identify each widget and determine which widget is selected. As
+/// required by the [Map] class, keys must be of consistent types
+/// and must be comparable. The ordering of the keys will determine the order
+/// of the widgets in the segmented control.
+///
+/// When the state of the segmented control changes, the widget changes [controller]'s
+/// value to the map key associated with the newly selected widget, causing all
+/// of its listeners to be notified.
+///
+/// {@tool sample}
+///
+/// ```dart
+/// class SegmentedControlExample extends StatefulWidget {
+///   @override
+///   State createState() => SegmentedControlExampleState();
+/// }
+///
+/// class SegmentedControlExampleState extends State<SegmentedControlExample> {
+///   final Map<int, Widget> children = const <int, Widget>{
+///     0: Text('Child 1'),
+///     1: Text('Child 2'),
+///   };
+///
+///   final ValueNotifier<int> controller = ValueNotifier<int>(0);
+///
+///   @override
+///   void initState() {
+///     super.initState();
+///     // Print a message whenever the currently selected widget changes.
+///     controller.addListener(() { print('currently selected ${controller.value}'); });
+///   }
+///
+///   @override
+///   Widget build(BuildContext context) {
+///     return Container(
+///       child: CupertinoSlidingSegmentedControl<int>(
+///         children: children,
+///         controller: controller,
+///       ),
+///     );
+///   }
+///
+///   @override
+///   void dispose() {
+///     super.dispose();
+///     controller.dispose();
+///   }
+/// }
+/// ```
+/// {@end-tool}
+///
+/// The [children] will be displayed in the order of the keys in the [Map].
+/// The height of the segmented control is determined by the height of the
+/// tallest widget provided as a value in the [Map] of [children].
+/// The width of each child in the segmented control will be equal to the width
+/// of widest child, unless the combined width of the children is wider than
+/// the available horizontal space. In this case, the available horizontal space
+/// is divided by the number of provided [children] to determine the width of
+/// each widget. The selection area for each of the widgets in the [Map] of
+/// [children] will then be expanded to fill the calculated space, so each
+/// widget will appear to have the same dimensions.
+///
+/// A segmented control may optionally be created with custom colors. The
+/// [thumbColor], [backgroundColor] arguments can be used to override the segmented
+/// control's colors from its defaults.
+///
+/// See also:
+///
+///  * <https://developer.apple.com/design/human-interface-guidelines/ios/controls/segmented-controls/>
+class CupertinoSlidingSegmentedControl<T> extends StatefulWidget {
```

Comment:
Assuming that we are keeping both, is there any code that can be shared between the two, or are there tests that could cover both?  I didn't look very closely, but just checking that you have tried to avoid code duplication as much as possible.

Reply:
The reason for keeping both is the API of `CupertinoSlidingSegmentedControl` is drastically different from `CupertinoSegmentedControl`. It's probably too big of a breaking change (and breaks every existing `CupertinoSegmentedControl`) to introduce in one go. Introducing it as a new widget will allow people to migrate as they feel comfortable. But yeah we will have to maintain both widgets with duplicated code for the time being now.

Reply:
Ah I see, makes sense.  Is it worth adding a `@deprecated` flag to CupertinoSegmentedControl now?

Reply:
IMO for now it doesn't hurt for them to coexist, as the comment you mentioned requested.